modules/common/react/lib/utils/components.ts

```diff
@@ -0,0 +1,174 @@
+  if (ref) {
+    if (typeof ref === 'function') {
+      ref(value);
+    } else {
+      // @ts-ignore Refs are readonly, but we can technically write to it without issue
+      ref.current = value;
+    }
+  }
+}
+
+/**
+ * This function will create a new forked ref out of two input Refs. This is useful for components
+ * that use `React.forwardRef`, but also need internal access to a Ref.
+ *
+ * This function is inspired by https://www.npmjs.com/package/@rooks/use-fork-ref
+ *
+ * @example
+ * React.forwardRef((props, ref) => {
+ *   // Returns a RefObject with a `current` property
+ *   const myRef = React.useRef(ref)
+ *
+ *   // Returns a forked Ref function to pass to an element.
+ *   // This forked ref will update both `myRef` and `ref` when React updates the element ref
+ *   const elementRef = useForkRef(ref, myRef)
+ *
+ *   useEffect(() => {
+ *     console.log(myRef.current) // `current` is the DOM instance
+ *     // `ref` might be null since it depends on if someone passed a `ref` to your component
+ *     // `elementRef` is a function and we cannot get a current value out of it
+ *   })
+ *
+ *   return <div ref={elementRef}/>
+ * })
+ */
+export function useForkRef<T>(ref1: React.Ref<T>, ref2: React.Ref<T>): (instance: T) => void {
+  return (value: T) => {
+    setRef(ref1, value);
+    setRef(ref2, value);
+  };
+}
+
+/**
+ * This function is the same as calling `React.useRef(null)` with the added benefit of extracting
+ * the type from the ref passed to it. This means you don't have to provide a generic to the
+ * `useRef` function
+ * @param ref The React ref passed from the `createComponent` factory function
+ */
+export function useComponentRef<T>(ref: React.Ref<T>) {
+  return React.useRef<T>(null);
+}
+
+/**
+ * Returns a model, or calls the model hook with config. Clever way around the conditional React
+ * hook ESLint rule.
+ * @param model A model, if provided
+ * @param config Config for a model
+ * @param modelHook A model hook that takes valid config
+ * @example
+ * const ContainerComponent = ({children, model, ...config}: ContainerProps) => {
+ *   const value = useDefaultModel(model, config, useContainerModel);
+ *
+ *   // ...
+ * }
+ */
+export function useDefaultModel<T, C>(
+  model: T | undefined,
+  config: C,
+  modelHook: (config: C) => T
+) {
+  return model || modelHook(config);
+}
+
+/**
+ * Returns a model, or returns a model context. Clever way around the conditional React hood ESLint
+ * rule
+ * @param model A model, if provided
+ * @param context The context of a model
+ * @example
+ * const SubComponent = ({children, model, ...elemProps}: SubComponentProps, ref, Element) => {
+ *   const {state, events} = useModelContext(model, SubComponentModelContext);
+ *
+ *   // ...
+ * }
+ */
+export function useModelContext<T>(context: React.Context<T>, model?: T) {
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  return model || React.useContext(context);
```

Comment:
hmm, so context would always be required for all sub-components, even if they were using the model? I'm not sure that's what we want? I'll have to think about it more.

As an example: Lets say I just want to do this:
```tsx
const tabsModel = useTabsModel(config);

<Tabs.List model={tabsModel}>
  <Tab model={tabModel}>One</Tab>
  <Tab model={tabModel}>Two</Tab>
  <Tab model={tabModel}>Three</Tab>
</Tabs.List>
```

Reply:
This adds the flexibility of optionally passing in a model. As far as context always being required... I'm not sure how that's a problem. If a React Context is not found it will return `null`. And in your example, the `React.useContext` would never be hit anyways, so in your example, no Context is created or used.